lib/fetcher/modulator.ts

```diff
@@ -0,0 +1,307 @@
+import { StateT } from "../state";
+import { getLoggerFor } from "../utils/";
+
+import type { StateFactory } from "../state";
+
+export type Notifier<Events, S> = {
+    [K in keyof Events]: (event: Events[K], state: S) => void;
+};
+
+/**
+ * Generic interface that represents a structure that ranks elements.
+ * Most common is a Priority Queue (heap like) that pops elements in order.
+ * An array is also a Ranker, without ordering.
+ */
+export interface Ranker<T> {
+    push(item: T): void;
+    pop(): T | undefined;
+}
+
+export type ModulartorEvents<T> = {
+    ready: Indexed<T>;
+};
+
+/**
+ * Modulator is a structure that only buffers elements and only handles elements
+ * when the factory is not paused and when not too many items are active at once.
```

Comment:
If I get it, the modulator is also some state unrelated to buffering elements.
And is used to record whether or not some things are (i) mutable (ii) immutable (iii) emittedd (iv) unemitted.
Just reading the code it is not clear what it means to be unemitted (for example), could you add a small comment?

Reply:
Addressed in https://github.com/rdf-connect/ldes-client/pull/94/commits/e72d20069219d1abf311859253db31b3e25cfc1b